[[ Bug 20644 ]] Fix calls to struct-returning obj-c methods

livecodeali opened the pull request:
Sometimes structs are returned from Objective-C methods in
registers instead of using the address provided as a parameter. In
the latter case `objc_msgSend_stret` should be used instead of
`objc_msgSend`. The rules for when this happens are architecture and
struct-size dependent and FFI calls will crash if the wrong form
is used. This patch updates the code to use `objc_msgSend_stret`
just in case the method returns a struct and
- on arm64, the struct size > 64 bytes
- on arm, the struct size > 4 bytes
- on x86_64, the struct size > 16 bytes
- on i386, the struct size > 8 bytes or a power of 2.

This could do with more comprehensive testing - but should do for now. It's not easy to find a simple struct in the obj-c libraries that is bigger than 16 bytes and therefore can test the use of `objc_msgSend_stret` on x86-64.

livecodeali committed:
[[ Bug 20644 ]] Fix calls to struct-returning obj-c methods
Sometimes structs are returned from Objective-C methods in
registers instead of using the address provided as a parameter. In
the latter case `objc_msgSend_stret` should be used instead of
`objc_msgSend`. The rules for when this happens are architecture and
struct-size dependent and FFI calls will crash if the wrong form
is used. This patch updates the code to use `objc_msgSend_stret`
just in case the method returns a struct and
- on arm64, the struct size > 64 bytes
- on arm, the struct size > 4 bytes
- on x86_64, the struct size > 16 bytes
- on i386, the struct size > 8 bytes or a power of 2.

runrevmark commented on libscript/src/script-execute.cpp:
```diff
@@ -289,22 +293,31 @@ class MCScriptForeignInvocation
             t_objc_msgSend = (void(*)())objc_msgSend;
 #elif defined(__X86_64__)
-        if (t_cif->rtype->type == FFI_TYPE_STRUCT)
+        if (t_cif->rtype->type == FFI_TYPE_STRUCT &&
```
Perhaps add a comment here about long double in the struct needing more investigation - just so we don't forget (or if someone wonders why some function using long doubles doesn't work!)

livecodeali committed:
[[ Bug 20644 ]] Additional comment and test
- Add a comment TODO to investigate structs with long doubles on
  x86-64 arch
- Add a test that takes objc_msgSend_stret path on x86-64 arch

runrevmark commented on tests/lcb/vm/interop-objc.lcb:
```diff
@@ -148,4 +148,36 @@ public handler TestReturnNSRange()
+	end if
+	
+	variable tRect as NSRect
+	put [0,0,0,0] into tRect
```
Hmmm - I'm not sure 0,0,0,0 is a good choice here as it is what you would get if some part of the process just happened to memory clear a target memory buffer before use - it would be better to use a rect which was composed of unlikely constant values (e.g. pi and e) which are *very* unlikely to be there.

livecodeali replied:
Ah yes, good point :-)

livecodeali committed:
[[ Bug 20644 ]] Use values for test unlikely to produce false positive

runrevmark commented:
@livecode-vulcan review ok dd81de3

livecode-vulcan commented:
💙 review by @runrevmark ok dd81de3473b6aa98d31bc9d617b8211b1e4d5c0e

livecode-vulcan commented:
😞 test failure dd81de3473b6aa98d31bc9d617b8211b1e4d5c0e

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/957)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/973)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/947)
* try-community-universal-ios-iphoneos11.1: [failure](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.1/builds/20)
* try-community-universal-ios-iphonesimulator11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.1/builds/20)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/951)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/974)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/957)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/973)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/956)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/966)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/988)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/952)

livecodeali committed:
[[ Bug 20644 ]] Arm64 does not have a _stret variant

runrevmark commented:
@livecode-vulcan review ok ea88681

livecode-vulcan commented:
💙 review by @runrevmark ok ea8868118c70e9ddbf632812704453d9b05fac6d

livecode-vulcan commented:
😎 test success ea8868118c70e9ddbf632812704453d9b05fac6d

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/964)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/980)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/954)
* try-community-universal-ios-iphoneos11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.1/builds/27)
* try-community-universal-ios-iphonesimulator11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.1/builds/27)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/958)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/981)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/964)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/980)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/963)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/973)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/995)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/959)
